scylla/src/transport/topology.rs

```diff
@@ -51,79 +57,95 @@ pub enum Strategy {
 
 impl TopologyReader {
     /// Creates new TopologyReader, which connects to known_peers in the background
-    pub fn new(known_peers: &[SocketAddr], connection_config: ConnectionConfig) -> Self {
-        let connect_port: u16 = known_peers
-            .first()
-            .expect("Tried to initialize TopologyReader with empty known_peers list!")
-            .port();
-
-        let mut control_connections: HashMap<SocketAddr, ConnectionKeeper> =
-            HashMap::with_capacity(known_peers.len());
-
-        for address in known_peers {
-            control_connections.insert(
-                *address,
-                ConnectionKeeper::new(*address, connection_config.clone(), None, None, None),
-            );
-        }
+    pub fn new(
+        known_peers: &[SocketAddr],
+        mut connection_config: ConnectionConfig,
+        server_event_sender: mpsc::Sender<Event>,
+    ) -> Self {
+        let control_connection_address = *known_peers
+            .last()
+            .expect("Tried to initialize TopologyReader with empty known_peers list!");
+
+        // setting event_sender field in connection config will cause control connection to
+        // - send REGISTER message to receive server events
+        // - send received events via server_event_sender
+        connection_config.event_sender = Some(server_event_sender);
+
+        let control_connection = ConnectionKeeper::new(
+            control_connection_address,
+            connection_config.clone(),
+            None,
+            None,
+            None,
+        );
 
         TopologyReader {
-            control_connections,
-            connect_port,
+            control_connection_address,
+            control_connection,
             connection_config,
+            known_peers: known_peers.into(),
         }
     }
 
     /// Fetches current topology info from the cluster
     pub async fn read_topology_info(&mut self) -> Result<TopologyInfo, QueryError> {
-        let result = self.fetch_topology_info().await;
+        let mut result = self.fetch_topology_info().await;
 
-        if let Ok(topology_info) = &result {
-            self.update_control_connections(topology_info);
-        }
+        // if fetching topology info on current control connection failed,
+        // try to fetch topology info from other known peer
+        for peer in self.known_peers.iter() {
```

Comment:
I think it makes sense to iterate over `known_peers` in random order to achieve better load-balancing. You could also filter out the `control_connection_address` node in this loop so that we don't try it twice.

Reply:
Done